### ISSUE TITLE
Simple shared memory model yields unfoldCTbranch: Not unique

### ISSUE THREAD

**Issue by thburghout** [TorXakis/TorXakis#900]
The following model consists of a process which reads an integer existing in a list synced with another process reading the same integer uncoditionally and outputting it through channel A.

```
 
TYPEDEF List_Int
    ::= NilList_Int
    |   List_Int { head :: Int; tail :: List_Int }
ENDDEF

CONSTDEF
    numbers :: List_Int ::= List_Int(1, List_Int(2, List_Int(3, NilList_Int)))
ENDDEF

FUNCDEF in (x :: Int; list :: List_Int) :: Bool
    ::= IF isNilList_Int(list) THEN False
        ELSE IF head(list) == x THEN True
        ELSE in(x, tail(list))
        FI FI
ENDDEF

PROCDEF ListState [Read :: Int] (state :: List_Int)
    ::= Read ? y [[ in(y, state) ]] >-> ListState [Read] (state)
ENDDEF

PROCDEF StartPs [A :: Int] ()
    ::= HIDE [ Read :: Int ] IN
        ListState [ Read ] (numbers)
    |[ Read ]|
        Do [A, Read] ()
    NI
ENDDEF

PROCDEF Do [A :: Int; Read :: Int] ()
    ::= Read ? y >-> A ! y >-> Do [A, Read] ()
ENDDEF

CHANDEF Chans ::=
    A :: Int
ENDDEF

MODELDEF Model ::=
    CHAN IN
    CHAN OUT A
    BEHAVIOUR StartPs [A] ()
ENDDEF
```

yields 

```
➜  prl-model-based-tester git:(master) ✗ torxakis model/memory.txs

TXS >>  TorXakis :: Model-Based Testing

TXS >>  txsserver starting: "localhost" : 33801
TXS >>  Solver "cvc4" initialized : cvc4 [1.6-prerelease]
TXS >>  TxsCore initialized
TXS >>  input files parsed:
TXS >>  model/memory.txs
TXS >> stepper Model
TXS >>  unfoldCTbranch: Not unique
TXS >>  Stepper started
TXS >> step 100
TXS >>  no state or deadlock
TXS >>  FAIL: No Output (Quiescence)
```

While I would expect the model to constantly yield either 1, 2, or 3.  The sample is greatly inspired by the shared memory example. I'm probably missing something here, but I cannot figure out what. 

It worth noting that not using shared memory (i.e. A ? x [[ in(x, numbers) ]] ) works fine. 

The same behaviour can be seen in both z3 as cvc4.

**Comment by pjljvandelaar** [comment]
I think you want:
```
PROCDEF Do [A :: Int; Read :: Int] ()
    ::= Read ? y | A ! y >-> Do [A, Read] ()
ENDDEF
```
Since in this case, the externally read value makes the internal/ hidden read value unique.

**Comment by thburghout** [comment]
Your proposed solution yields an Undefined Entity for y communicated over A. ~~Could this be a bug in the latest linux version (0.7) fixed in the latest releases (0.8+) ?~~

I am using 0.7.0 on linux. I tested it on 0.8.1 as well.

**Comment by thburghout** [comment]
If I replace `A ! y` with `A ! 20`, which is not the desired behaviour; since I want a value from the state list. It yields another runtime error.

```
TXS >> stepper Model
TXS >>  Stepper started
TXS >> step 4
TXS >>  .....1: Act { { ( A, [ 20 ] ) } }
TXS >>  after: cannot find unique value for hidden variables
TXS >>  stepN - cannot do action
TXS >>  FAIL: Act { { ( A, [ 20 ] ) } }
```

Are there some resources available on the topic of the uniqueness of values? What are the specific constraints here? Why must it find a unique value? Any value from the state [1,2,3] should be fine.

**Comment by pjljvandelaar** [comment]
Sorry i was too fast:
'Read ? y1| A ? y2 [[y1 ==y2]] >-> Do [A, Read] ()'

TorXakis requires that all hidden communication is unique to ensure the number of states to be considered remains finite. The wiki should contain this info. Please complain / submit an issue when you can't find it or it is not sufficient for you.
I can't promise we will solve it quickly, but it is archived....

**Comment by pjljvandelaar** [comment]
BTW
Why don't you simplify
```
PROCDEF ListState [Read :: Int] (state :: List_Int)
    ::= Read ? y [[ in(y, state) ]] >-> ListState [Read] (state)
ENDDEF

PROCDEF StartPs [A :: Int] ()
    ::= HIDE [ Read :: Int ] IN
        ListState [ Read ] (numbers)
    |[ Read ]|
        Do [A, Read] ()
    NI
ENDDEF

PROCDEF Do [A :: Int; Read :: Int] ()
    ::= Read ? y >-> A ! y >-> Do [A, Read] ()
ENDDEF
```
to just
```
PROCDEF StartPs [A :: Int] ()
    ::= A ? x [[ in(x,numbers) ]] >-> StartPs [A]()
ENDDEF
```

**Comment by thburghout** [comment]
It was part of a larger model, where I need to synchronise state updates between processes.  There I ran into the lack of uniqueness. When recreating the issue in a smaller model I found that the Write did not cause issues; hence I left it out of the example in this issue.

Your suggestion `Read ? y1| A ? y2 [[y1 ==y2]] >-> Do [A, Read] ()` resolved the issue in the example, thanks! I'm integrating it into the larger model right now.

I will close the issue for now.
